### PR TITLE
Output a Symmetric from the Hessians

### PR DESCRIPTION
**Pull request by ChrisRackauckas**

### CODE REVIEW

**Comment by tpapp** [comment]
I think this is preferable, but note that this relies on all callers being written in a generic way (accepting `<:AbstractMatrix` instead of `Matrix`). Which is how it should be, but you may want to test it in typical applications.

**Comment by ChrisRackauckas** [comment]
@pkofod is Optim.jl fine with this?

**Comment by pkofod** [comment]
I’m not sure. This will only cause problems if you use the returned or object, and I don’t think we do. I guess the safe thing to do is then just to also input a matrix wrapped in symmetric right ?

But remind me again, why are we doing this if the triangle is already being copied?

**Comment by tpapp** [comment]
> But remind me again, why are we doing this if the triangle is already being copied?

Generally, there are operations on symmetric (`::Array` etc) matrices that should, but don't preserve symmetry (numerically), while for `::Symmetric` they could do this automatically. Not many are supported as primitives at the moment though.

**Comment by pkofod** [comment]
Okay. I guess I just prefer H in H out in these types of functions and then it and then people can decorate their stuff with symmetric as they like. I mean if they specifically want the symmetry-preserving (but potentially different in other aspects ) algorithms, then they call symmetric on it themselves. But I’m not really religious on this. As long as the triangle is copied.

**Comment by tpapp** [comment]
> then people can decorate their stuff with symmetric as they like

In general, this assumes an understanding of the mathematical properties of the algorithm that may change, etc. I think the function that produces a value and can ensure a particular property should be the one signaling it.

> As long as the triangle is copied.

My understanding is that this does **not** happen (or meaningless to rely on) for `Symmetric`.

**Comment by pkofod** [comment]
> My understanding is that this does **not** happen (or meaningless to rely on) for `Symmetric`.

It does for the input `H` here, and that was specifically what I was referring to. For a mutating interface where you specifically pass in `H` it should either very clearly be documented that it only operates on the upper triangle or it should fill out everything.

**Comment by tpapp** [comment]
@pkofod: I think you are right. I missed the context, specifically that this is already a mutating function. I was thinking more along the lines of
```julia
julia> f(x) = Diagonal(x) # hessian of sum(x.^3 ./ 6)
f (generic function with 1 method)

julia> f!(H, x) = H .= f(x)
f! (generic function with 1 method)

julia> H = randn(3, 3)
3×3 Array{Float64,2}:
  0.974444  0.0354839  -0.241943
  2.32276   1.94886     0.657737
 -1.08576   1.24153    -0.549359

julia> f!(H, ones(3))
3×3 Array{Float64,2}:
 1.0  0.0  0.0
 0.0  1.0  0.0
 0.0  0.0  1.0
```
where the API is `f` but someone else in `f!` is using it in a mutating way.

So as given, this PR may be incorrect since if the caller gives a `H::Matrix`, they should expect that to be filled out.

**Comment by pkofod** [comment]
👍

File: src/hessians.jl [hunk 1]
```diff
@@ -34,7 +34,7 @@ function finite_difference_hessian(
     _H = false .* x .* x'
     _H isa SMatrix ? H = MArray(_H) : H = _H
     finite_difference_hessian!(H, f, x, cache; relstep=relstep, absstep=absstep)
-    _H isa SMatrix ? SArray(H) : H
+    _H isa SMatrix ? Symmetric(SArray(H)) : Symmetric(H)
```
Comment (tpapp) [inline src/hessians.jl]:
```suggestion
    Symmetric(_H isa SMatrix ? SArray(H) : H)
```

**Pull request merged**